In [1]:
# Importando pandas
import pandas as pd

# Carregando os dados
df = pd.read_csv('C:/my_workspace/triple_ten/project2/data/credit_scoring_eng.csv')

In [2]:
# Chamando .info() para olhar os dados
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


`Dados possuem 21525 entradas em 12 colunas.`

`Colunas ['days_employed'] e ['total_income'] apresentam valores nulos.`

In [3]:
# chamando .decribe
print(df.describe)

<bound method NDFrame.describe of        children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

     

In [4]:
# chamando .head()
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

`Overview`
- `Na coluna ['days_employed'] foram encontrados valores negativos`
- `Na coluna ['education'] temos dados em upper case e também em lower case`
- `Coluna ['total_income'] apresenta valores NaN`


In [5]:
# chamando .tail()
df.tail(15)

children  days_employed  dob_years            education  education_id  \
21510         2            NaN         28  secondary education             1   
21511         0    -612.569129         29    bachelor's degree             0   
21512         0    -165.377752         26    bachelor's degree             0   
21513         0   -1166.216789         35  secondary education             1   
21514         0    -280.469996         27         some college             2   
21515         1    -467.685130         28  secondary education             1   
21516         0    -914.391429         42    bachelor's degree             0   
21517         0    -404.679034         42    bachelor's degree             0   
21518         0  373995.710838         59  SECONDARY EDUCATION             1   
21519         1   -2351.431934         37      graduate degree             4   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21510            married                 0      F    employee     0   
21511  civil partnership                 1      F    employee     1   
21512          unmarried                 4      M    business     0   
21513            married                 0      F    employee     0   
21514          unmarried                 4      M    business     0   
21515            married                 0      F    employee     1   
21516            married                 0      F    business     0   
21517  civil partnership                 1      F    business     0   
21518            married                 0      F     retiree     0   
21519           divorced                 3      M    employee     0   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                           purpose  
21510           NaN                      car purchase  
21511     22410.956   buying property for renting out  
21512     23568.233  to get a supplementary education  
21513     40157.783             purchase of the house  
21514     56958.145               building a property  
21515     17517.812                to become educated  
21516     51649.244          purchase of my own house  
21517     28489.529                 buying my own car  
21518     24618.344                 purchase of a car  
21519     18551.846        buy commercial real estate  
21520     35966.698              housing transactions  
21521     24959.969                 purchase of a car  
21522     14347.610                          property  
21523     39054.888                 buying my own car  
21524     13127.587                      to buy a car

`Overview`
- `Na coluna [days_employed] foram encontrados além de valores negativos, também valores NaN.`


`Como encontramos valores NaN em duas colunas numa primeira análise, teremos que tratar esses dados.`

`Transformaremos a coluna ['education'] em lower case na sequência e avaliaremos quantos valores únicos temos na coluna ['purpose'].`


`Contando valores ausentes:`

In [6]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

`Confirmação de um grande volume de dados ausentes nas colunas ['days_employed'] e ['total_income']`

`Visualizando a quantidade de dados duplicados:`

In [7]:
df.duplicated().sum()

54

`Temos 54 dados duplicados, que trataremos mais tarde, por hora, vamos focar nos valores ausentes.`

`Examinando valores ausentes:`

In [8]:
# Vamos aplicar várias condições para filtrar dados e observar o número de linhas na tabela filtrada.

# valores ausentes nas colunas days_employed e total income

missing_values_total_income = df['total_income'].isna().sum()
print(f'{missing_values_total_income} valores ausentes na coluna total_income')

missing_values_days_employed = df['days_employed'].isna().sum()
print(f'{missing_values_days_employed} valores ausentes na coluna days_employed')

2174 valores ausentes na coluna total_income
2174 valores ausentes na coluna days_employed


`Uma quantidade elevada de dados ausentes, veremos em porcentagem  o quanto representa do dataframe.`

In [9]:
# df.shape para verificar quantidades de linhas 
df.shape

# salvando em uma variável para calcular a porcentagem
number_of_lines = len(df)
number_of_lines

# calculando a porcentagem de dados ausentes 
percentage_of_missing_values = missing_values_total_income / number_of_lines
print('A porcentagem de dados ausentes é de:', round(percentage_of_missing_values * 100),'%')

A porcentagem de dados ausentes é de: 10 %


`Percebemos aqui que temos 10% de dados ausentes na coluna ['days_employed'] e valor igual na coluna ['total_income'].`

`Como é um porcentagem relevante, não iremos excluir, mas sim iremos preencher com valores relevantes.`

In [10]:
# Vamos investigar clientes que não possuem dados sobre as características identificadas e a coluna com os valores ausentes

both_columns = df[(df['total_income'].isna()) & (df['days_employed'].isna())]
both_columns

# descobrimos aqui que todos os dados ausentes se apresentam em ambas as colunas ao mesmo tempo

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [11]:
# Verificando a distribuição por 'education'
grouping_for_analysis_education = both_columns.groupby('education').count()
grouping_for_analysis_education

children  days_employed  dob_years  education_id  \
education                                                               
BACHELOR'S DEGREE          23              0         23            23   
Bachelor's Degree          25              0         25            25   
PRIMARY EDUCATION           1              0          1             1   
Primary Education           1              0          1             1   
SECONDARY EDUCATION        67              0         67            67   
SOME COLLEGE                7              0          7             7   
Secondary Education        65              0         65            65   
Some College                7              0          7             7   
bachelor's degree         496              0        496           496   
primary education          19              0         19            19   
secondary education      1408              0       1408          1408   
some college               55              0         55            55   

                     family_status  family_status_id  gender  income_type  \
education                                                                   
BACHELOR'S DEGREE               23                23      23           23   
Bachelor's Degree               25                25      25           25   
PRIMARY EDUCATION                1                 1       1            1   
Primary Education                1                 1       1            1   
SECONDARY EDUCATION             67                67      67           67   
SOME COLLEGE                     7                 7       7            7   
Secondary Education             65                65      65           65   
Some College                     7                 7       7            7   
bachelor's degree              496               496     496          496   
primary education               19                19      19           19   
secondary education           1408              1408    1408         1408   
some college                    55                55      55           55   

                     debt  total_income  purpose  
education                                         
BACHELOR'S DEGREE      23             0       23  
Bachelor's Degree      25             0       25  
PRIMARY EDUCATION       1             0        1  
Primary Education       1             0        1  
SECONDARY EDUCATION    67             0       67  
SOME COLLEGE            7             0        7  
Secondary Education    65             0       65  
Some College            7             0        7  
bachelor's degree     496             0      496  
primary education      19             0       19  
secondary education  1408             0     1408  
some college           55             0       55

`Não existe padrão olhando por agrupamento na coluna ['education']. Vamos olhar a relação entre ['days_employed'] e  ['income_type'] mais abaixo.`

In [12]:
# Verificando a distribuição por 'days_employed'
grouping_for_analysis_days_employed = both_columns.groupby('days_employed').count()
grouping_for_analysis_days_employed

Empty DataFrame
Columns: [children, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [13]:
# Verificando a distribuição por 'total_income'
grouping_for_analysis_total_income = both_columns.groupby('total_income').count()
grouping_for_analysis_total_income

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, purpose]
Index: []

`Existe uma relação entre a coluna ['days_employed'] ser 0 (zero) e a ['total_income'] para os nossos valores ausentes.`

`Podemos considerar em primeira instância que essas pessoas estão desempregadas ou momentaneamente sem renda.`

In [14]:
# Vamos ver todos os valores na coluna de educação para verificar se e quais grafias precisarão ser corrigidas
df['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

`Transformando todos em lower case`

In [15]:
# Corrija os registros, se necessário
df['education'] = df['education'].str.lower()

In [16]:
# Verificando todos os valores na coluna para ter certeza de que os corrigimos
df['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

`Dados na coluna ['education'] foram corrigidos.`

In [17]:
# Vamos ver a distribuição de valores na coluna `children`
df['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

`Existem dados bem estranhos na coluna ['children']: -1 e 20`

`Iremos contar as aparições desses dados`

In [18]:
# verificanda a aparição de 20 filhos

twenty_children = (df['children'] == 20).sum()
print(f"A quantidade de vezes que aparece o valor 20 na coluna ['children'] é: {twenty_children}")

print((twenty_children / number_of_lines) * 100)

A quantidade de vezes que aparece o valor 20 na coluna ['children'] é: 76
0.3530778164924506


`Valor irrelevante, em torno de 0,35% de todo o dataframe.`

`Buscando dados na internet sobre a média mundial de filhos em nossa década atual, que é de 2,1. Iremos alterar para 2, sugerindo que deve ter sido um erro de input.`

In [19]:
# verificanda a aparição de -1 filhos

minus_one_child = (df['children'] == -1).sum()
print(f"A quantidade de vezes que aparece o valor -1 na coluna ['children'] é: {minus_one_child}")

print((minus_one_child / number_of_lines) * 100)

A quantidade de vezes que aparece o valor -1 na coluna ['children'] é: 47
0.2183507549361208


`Valor também irrelevante, em torno de 0,21% de todo o dataframe`

`Supomos aqui, que foi novamente um erro de digitação e transformaremos o valor de -1 em 1.`

In [20]:
# [corrija os dados com base na sua decisão]
df['children'] = df['children'].replace(20, 2)

df['children'] = df['children'].replace(-1, 1)

In [21]:
# Verificar a coluna `children` novamente para ter certeza de que está tudo corrigido
df['children'].unique()

array([1, 0, 3, 2, 4, 5])

`Dados da coluna ['children'] corrigidos.`

In [22]:
# Encontre dados problemáticos em `days_employed`, se existirem, e calcule a porcentagem
df['days_employed'].unique()

array([-8437.67302776, -4024.80375385, -5623.42261023, ...,
       -2113.3468877 , -3112.4817052 , -1984.50758853])

In [23]:
# são muitos valores, vamos olhar mais de perto
count_the_negatives = (df['days_employed'] < 0).sum()
count_the_negatives

15906

In [24]:
# representatividade dos valores negativos no dataframe
print((count_the_negatives / number_of_lines) * 100)

73.89547038327527


`Representam +73% do DataFrame.`

`Como não podemos ignorar esses valores, assumiremos que houve problemas técnicos e passaremos os valores para positivo, pois não existe a possibilidade de uma pessoa trabalhar dias negativos.`

In [25]:
# Aborde os valores problemáticos, se existirem
df['days_employed'] = df['days_employed'].apply(lambda x: x if x >= 0 else -x)

In [26]:
# Verifique o resultado - certifique-se de que está corrigido
count_the_negatives = (df['days_employed'] < 0).sum()
count_the_negatives

0

In [27]:
# Verifique o `dob_years` para valores suspeitos e conte a porcentagem
df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [28]:
# Contando quantos valores 0 existem em '[dob_years']
impossible_ge = (df['dob_years'] == 0).sum()
impossible_ge

101

`Existem 101 valores problemáticos na coluna ['dob_years'] e faremos uma verificação da mediana, pois a média é de 43.29 informada pelo .describe()`

In [29]:
# Calculando a mediana
impossible_age_median = df['dob_years'].median()
impossible_age_median

42.0

`Como diferença entre a média e a mediana foi pequena, isso indica uma distribuição aproximadamente simétrica. Faremos a substituição dos valores 0 na coluna ['dob_years'] por 42.`

In [30]:
# Resolva os problemas na coluna `dob_years`, se existirem
df['dob_years'] = df['dob_years'].replace(0, 42)

In [31]:
# Verifique o resultado - certifique-se de que está corrigido
df['dob_years'].describe()

count    21525.000000
mean        43.490453
std         12.218595
min         19.000000
25%         34.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [32]:
# Verificando novamente
impossible_ge = (df['dob_years'] == 0).sum()
impossible_ge

0

`Problemas resolvidos na coluna ['dob_years']`

In [33]:
# Vamos ver os valores da coluna
df['family_status'].describe()

count       21525
unique          5
top       married
freq        12380
Name: family_status, dtype: object

In [34]:
# Aborde os valores problemáticos em `family_status`, se eles existirem
df['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In [35]:
df['family_status'].isna().sum()

0

In [36]:
# Verifique o resultado - certifique-se de que está corrigido

`Não existem dados com problemas na coluna ['family_status'] e nem valores ausentes.`

In [37]:
# Vamos ver os valores na coluna
df['gender'].describe()

count     21525
unique        3
top           F
freq      14236
Name: gender, dtype: object

In [38]:
df['gender'].isna().sum()

0

In [39]:
df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [40]:
gender_xna = (df['gender'] == 'XNA').sum()
gender_xna

1

`Encontrado um valor 'XNA' apenas. Não idetificamos o que significa. Olharemos todos os dados da linha onde se encontra esse 'XNA'`

In [41]:
# Localizando XNA
find_xna = df.loc[df['gender'] == 'XNA']
find_xna

children  days_employed  dob_years     education  education_id  \
10701         0    2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

`Como não há uma informação que possamos definir qual o 'gender de 'XNA', substituiremos pelo 'gender' de maior número no DataFrame.`

In [42]:
# Verifique o resultado - certifique-se de que está corrigido
count_male = (df['gender'] == 'F').sum()
count_male

14236

`'M' está em maior número.`

In [43]:
# Substituindo 'XNA' por 'M'
df['gender'] = df['gender'].replace('XNA', 'M')

In [44]:
# Certificando
df['gender'].unique()

array(['F', 'M'], dtype=object)

In [45]:
# Vamos ver os valores na coluna
df['income_type'].describe()

count        21525
unique           8
top       employee
freq         11119
Name: income_type, dtype: object

In [46]:
df['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [47]:
df['income_type'].isna().sum()

0

`Sem problemas na coluna ['income_type']`

In [48]:
# Verificar duplicados
df.duplicated().sum()

72

In [49]:
df_duplicados = df.duplicated()
print(df_duplicados)

0        False
1        False
2        False
3        False
4        False
         ...  
21520    False
21521    False
21522    False
21523    False
21524    False
Length: 21525, dtype: bool


In [50]:
# Corrija os duplicados, se existirem
df = df.drop_duplicates().reset_index()

In [51]:
# Última verificação se temos duplicados
df.duplicated().sum()

0

`Não existem mais dados duplicados aparentes.`

In [52]:
#Verifique o tamanho do conjunto de dados que você tem agora após suas primeiras manipulações com ele
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             21453 non-null  int64  
 1   children          21453 non-null  int64  
 2   days_employed     19351 non-null  float64
 3   dob_years         21453 non-null  int64  
 4   education         21453 non-null  object 
 5   education_id      21453 non-null  int64  
 6   family_status     21453 non-null  object 
 7   family_status_id  21453 non-null  int64  
 8   gender            21453 non-null  object 
 9   income_type       21453 non-null  object 
 10  debt              21453 non-null  int64  
 11  total_income      19351 non-null  float64
 12  purpose           21453 non-null  object 
dtypes: float64(2), int64(6), object(5)
memory usage: 2.1+ MB


`Temos agora uma nova coluna, 13 no total, agora ela recebe o 'index'.`

In [53]:
# Diferença de linhas do início da Análise até agora
number_of_lines_after_first_modification = len(df)
number_of_lines_after_first_modification

line_difference = number_of_lines - number_of_lines_after_first_modification
line_difference

72

`Fizemos diversos ajustes e agora teremos que trabalhar os valores ausentes.`

# Trabalhando com valores ausentes

In [54]:
family_dict = {
    '0': 'married',
    '1': 'civil partnership',
    '2': 'widow / widower',
    '3': 'divorced',
    '4': 'unmarried',
}

In [55]:
education_dict = {
    '0': "bachelor's degree",
    '1': 'secondary education',
    '2': 'some college',
    '3': 'primary education',
    '4': 'graduate degree',
}

### Restaurar valores ausentes em `total_income`

In [56]:
df['dob_years'].describe()

count    21453.000000
mean        43.469025
std         12.214162
min         19.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [57]:
# Vamos escrever uma função que calcule a categoria de idade
def age_category(age):
    if age < 20:
        return 'Adolescent'
    elif age < 35:
        return 'Young adult'
    elif age < 50:
        return 'Adult'
    elif age < 65:
        return 'Middle-aged Adult'
    else:
        return 'Senior citizen / Elderly'

In [58]:
# Teste se a função funciona
for age in df['dob_years']:
    category = age_category(age)
    print(f'A categoria para a idade de {age} anos é: {category}')

A categoria para a idade de 42 anos é: Adult
A categoria para a idade de 36 anos é: Adult
A categoria para a idade de 33 anos é: Young adult
A categoria para a idade de 32 anos é: Young adult
A categoria para a idade de 53 anos é: Middle-aged Adult
A categoria para a idade de 27 anos é: Young adult
A categoria para a idade de 43 anos é: Adult
A categoria para a idade de 50 anos é: Middle-aged Adult
A categoria para a idade de 35 anos é: Adult
A categoria para a idade de 41 anos é: Adult
A categoria para a idade de 36 anos é: Adult
A categoria para a idade de 40 anos é: Adult
A categoria para a idade de 65 anos é: Senior citizen / Elderly
A categoria para a idade de 54 anos é: Middle-aged Adult
A categoria para a idade de 56 anos é: Middle-aged Adult
A categoria para a idade de 26 anos é: Young adult
A categoria para a idade de 35 anos é: Adult
A categoria para a idade de 33 anos é: Young adult
A categoria para a idade de 53 anos é: Middle-aged Adult
A categoria para a idade de 48 anos 

In [59]:
# Criar coluna nova com base na função
df['age_category'] = df['dob_years'].apply(age_category)

In [60]:
# Verificando a coluna criada
df.columns

Index(['index', 'children', 'days_employed', 'dob_years', 'education',
       'education_id', 'family_status', 'family_status_id', 'gender',
       'income_type', 'debt', 'total_income', 'purpose', 'age_category'],
      dtype='object')

In [61]:
# Verificar como ficaram os valores na nova coluna
df['age_category'].unique()

array(['Adult', 'Young adult', 'Middle-aged Adult',
       'Senior citizen / Elderly', 'Adolescent'], dtype=object)

In [62]:
# Verificando a distribuição
count_adolescents= df['age_category'].value_counts()['Adolescent']
count_adolescents

adolescent_rows = df[df['dob_years'] < 20]
adolescent_rows

index  children  days_employed  dob_years            education  \
766      766         0     796.983636         19         some college   
1981    1981         0     724.492610         19  secondary education   
2725    2725         0     115.876036         19  secondary education   
4050    4052         0            NaN         19  secondary education   
4096    4098         0     111.762790         19  secondary education   
5558    5563         0     885.268574         19         some college   
8307    8316         0     556.088508         19  secondary education   
9208    9218         0     322.024011         19  secondary education   
10220  10235         0     793.358581         19  secondary education   
11730  11749         0     757.612624         19  secondary education   
12027  12047         0    1020.183130         19  secondary education   
12998  13021         0     695.968951         19         some college   
15623  15659         0     948.225743         19         some college   
20165  20230         0     509.969922         19         some college   

       education_id      family_status  family_status_id gender  \
766               2            married                 0      F   
1981              1          unmarried                 4      F   
2725              1          unmarried                 4      F   
4050              1  civil partnership                 1      M   
4096              1  civil partnership                 1      M   
5558              2          unmarried                 4      M   
8307              1          unmarried                 4      F   
9208              1  civil partnership                 1      F   
10220             1            married                 0      F   
11730             1            married                 0      F   
12027             1          unmarried                 4      M   
12998             2          unmarried                 4      F   
15623             2  civil partnership                 1      F   
20165             2          unmarried                 4      F   

         income_type  debt  total_income  \
766         employee     0     12929.944   
1981        business     0      9459.851   
2725        business     0     21114.762   
4050        employee     0           NaN   
4096        business     0     14700.281   
5558        employee     0     14575.717   
8307        employee     0     10963.857   
9208        business     0     16588.237   
10220       employee     0     21009.404   
11730       business     0     19321.445   
12027       employee     1     14934.901   
12998       business     0     26443.044   
15623       business     0     26753.823   
20165  civil servant     0     12125.986   

                                        purpose age_category  
766             buying property for renting out   Adolescent  
1981                               to own a car   Adolescent  
2725   transactions with commercial real estate   Adolescent  
4050                           having a wedding   Adolescent  
4096                           having a wedding   Adolescent  
5558            buying property for renting out   Adolescent  
8307                       university education   Adolescent  
9208                          to have a wedding   Adolescent  
10220                        housing renovation   Adolescent  
11730                           buy real estate   Adolescent  
12027          transactions with my real estate   Adolescent  
12998                                      cars   Adolescent  
15623                  buying a second-hand car   Adolescent  
20165                buy commercial real estate   Adolescent

In [63]:
# Crie uma tabela sem valores ausentes e imprima algumas de suas linhas para garantir que ela fique boa
df_without_nan = df.dropna(subset=['total_income'])
df_without_nan.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19351 entries, 0 to 21452
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             19351 non-null  int64  
 1   children          19351 non-null  int64  
 2   days_employed     19351 non-null  float64
 3   dob_years         19351 non-null  int64  
 4   education         19351 non-null  object 
 5   education_id      19351 non-null  int64  
 6   family_status     19351 non-null  object 
 7   family_status_id  19351 non-null  int64  
 8   gender            19351 non-null  object 
 9   income_type       19351 non-null  object 
 10  debt              19351 non-null  int64  
 11  total_income      19351 non-null  float64
 12  purpose           19351 non-null  object 
 13  age_category      19351 non-null  object 
dtypes: float64(2), int64(6), object(6)
memory usage: 2.2+ MB


In [64]:
# Veja os valores médios de renda com base nos seus fatores identificados
df_without_nan.groupby('age_category')['total_income'].mean()

age_category
Adolescent                  16993.942462
Adult                       28544.058661
Middle-aged Adult           25313.124503
Senior citizen / Elderly    21542.650450
Young adult                 26692.462524
Name: total_income, dtype: float64

In [65]:
# Veja os valores medianos de renda com base nos seus fatores identificados
df_without_nan.groupby('age_category')['total_income'].median()

age_category
Adolescent                  14934.9010
Adult                       24790.9655
Middle-aged Adult           21830.2500
Senior citizen / Elderly    18471.3910
Young adult                 23427.9560
Name: total_income, dtype: float64

In [66]:
income_stats_by_education = df_without_nan.groupby(['age_category', 'education'])['total_income'].agg(['median', 'mean'])
income_stats_by_education

median          mean
age_category             education                                    
Adolescent               secondary education  15761.5690  16011.592250
                         some college         14575.7170  18565.702800
Adult                    bachelor's degree    29783.4780  35435.669510
                         graduate degree      18551.8460  22715.308000
                         primary education    21060.6930  23319.755429
                         secondary education  23036.7350  26104.983856
                         some college         30526.5090  33695.646036
Middle-aged Adult        bachelor's degree    27429.4230  32982.335874
                         graduate degree      41906.9125  41906.912500
                         primary education    17657.4995  18676.032750
                         secondary education  20807.0725  23588.439930
                         some college         24360.4600  28287.081922
Senior citizen / Elderly bachelor's degree    23353.5110  28177.463470
                         graduate degree      15800.3990  15800.399000
                         primary education    15354.3830  17332.906000
                         secondary education  17943.5555  20273.532997
                         some college         27938.4350  26241.578462
Young adult              bachelor's degree    27044.4440  31159.471620
                         primary education    20705.8340  23971.452780
                         secondary education  21935.2640  24322.663546
                         some college         23787.6330  27145.181245

In [67]:
income_stats_by_income_type = df_without_nan.groupby(['age_category', 'income_type'])['total_income'].agg(['median', 'mean'])
income_stats_by_income_type

median          mean
age_category             income_type                                          
Adolescent               business                     19321.4450  19197.349000
                         civil servant                12125.9860  12125.986000
                         employee                     14575.7170  14882.764600
Adult                    business                     28837.2555  33929.898979
                         civil servant                24431.3170  28176.906616
                         employee                     23156.8480  26242.858857
                         paternity / maternity leave   8612.6610   8612.661000
                         retiree                      21972.2140  26054.145364
                         unemployed                   32435.6020  32435.602000
Middle-aged Adult        business                     27499.3355  32382.127995
                         civil servant                23068.1320  26502.720719
                         employee                     22644.5210  26251.676702
                         retiree                      19370.9975  22101.126638
Senior citizen / Elderly business                     26948.7450  31668.009026
                         civil servant                26089.6870  26921.808294
                         employee                     24660.9010  26139.495246
                         retiree                      17645.5060  20448.866997
Young adult              business                     26435.9690  30393.683928
                         civil servant                24290.0170  26699.399233
                         employee                     22236.4760  25009.843452
                         entrepreneur                 79866.1030  79866.103000
                         retiree                      15083.2080  18940.138533
                         student                      15712.2600  15712.260000
                         unemployed                    9593.1190   9593.119000

In [68]:
income_stats_by_gender = df_without_nan.groupby(['age_category', 'gender'])['total_income'].agg(['median', 'mean'])
income_stats_by_gender

median          mean
age_category             gender                          
Adolescent               F       17954.8410  17671.035300
                         M       14700.2810  14736.966333
Adult                    F       22669.4700  25986.543516
                         M       28395.9820  33176.268334
Middle-aged Adult        F       20822.0100  24032.492005
                         M       25188.3725  28897.682451
Senior citizen / Elderly F       18031.1620  20945.112382
                         M       19557.0440  23451.707969
Young adult              F       21341.9120  24180.916076
                         M       26768.9325  30197.231745

`Verificamos que os valores para a média são relativamente maiores dos que os valores encontrados com a mediana. Decidimos aqui usar, então a mediana para substituir os valores Nan. As colunas ['education'], ['gender'] e ['income_type'] são as que mais definem a renda.`

`Iremos usar a coluna ['gender'] para definir a renda total.`

In [69]:
avg_male = df_without_nan[ df_without_nan['gender'] == 'M']['total_income'].median()
avg_male

26838.552000000003

In [70]:
avg_female = df_without_nan[ df_without_nan['gender'] == 'F']['total_income'].median()
avg_female

21464.845

In [71]:
def filling_nan(avg_male, avg_female):
    
    df.loc[(df['total_income'].isna()) & (df['gender'] == 'M'), 'total_income'] = avg_male

    df.loc[(df['total_income'].isna()) & (df['gender'] == 'F'), 'total_income'] = avg_female

In [72]:
filling_nan(avg_male, avg_female)

In [73]:
# Verifique se funciona
df.isna().sum()

index                  0
children               0
days_employed       2102
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_category           0
dtype: int64

In [74]:
# Verificar o número de entradas nas colunas
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             21453 non-null  int64  
 1   children          21453 non-null  int64  
 2   days_employed     19351 non-null  float64
 3   dob_years         21453 non-null  int64  
 4   education         21453 non-null  object 
 5   education_id      21453 non-null  int64  
 6   family_status     21453 non-null  object 
 7   family_status_id  21453 non-null  int64  
 8   gender            21453 non-null  object 
 9   income_type       21453 non-null  object 
 10  debt              21453 non-null  int64  
 11  total_income      21453 non-null  float64
 12  purpose           21453 non-null  object 
 13  age_category      21453 non-null  object 
dtypes: float64(2), int64(6), object(6)
memory usage: 2.3+ MB


`Todos os dados corrigidos na coluna ['total_income'].`

###  Restaurar valores em `days_employed`

In [75]:
df.columns

Index(['index', 'children', 'days_employed', 'dob_years', 'education',
       'education_id', 'family_status', 'family_status_id', 'gender',
       'income_type', 'debt', 'total_income', 'purpose', 'age_category'],
      dtype='object')

In [76]:
df.head(15)

index  children  days_employed  dob_years            education  \
0       0         1    8437.673028         42    bachelor's degree   
1       1         1    4024.803754         36  secondary education   
2       2         0    5623.422610         33  secondary education   
3       3         3    4124.747207         32  secondary education   
4       4         0  340266.072047         53  secondary education   
5       5         0     926.185831         27    bachelor's degree   
6       6         0    2879.202052         43    bachelor's degree   
7       7         0     152.779569         50  secondary education   
8       8         2    6929.865299         35    bachelor's degree   
9       9         0    2188.756445         41  secondary education   
10     10         2    4171.483647         36    bachelor's degree   
11     11         0     792.701887         40  secondary education   
12     12         0            NaN         65  secondary education   
13     13         0    1846.641941         54         some college   
14     14         0    1844.956182         56    bachelor's degree   

    education_id      family_status  family_status_id gender income_type  \
0              0            married                 0      F    employee   
1              1            married                 0      F    employee   
2              1            married                 0      M    employee   
3              1            married                 0      M    employee   
4              1  civil partnership                 1      F     retiree   
5              0  civil partnership                 1      M    business   
6              0            married                 0      F    business   
7              1            married                 0      M    employee   
8              0  civil partnership                 1      F    employee   
9              1            married                 0      M    employee   
10             0            married                 0      M    business   
11             1            married                 0      F    employee   
12             1  civil partnership                 1      M     retiree   
13             2            married                 0      F    employee   
14             0  civil partnership                 1      F    business   

    debt  total_income                              purpose  \
0      0     40620.102                purchase of the house   
1      0     17932.802                         car purchase   
2      0     23341.752                purchase of the house   
3      0     42820.568              supplementary education   
4      0     25378.572                    to have a wedding   
5      0     40922.170                purchase of the house   
6      0     38484.156                 housing transactions   
7      0     21731.829                            education   
8      0     15337.093                     having a wedding   
9      0     23108.150  purchase of the house for my family   
10     0     18230.959                      buy real estate   
11     0     12331.077           buy commercial real estate   
12     0     26838.552                    to have a wedding   
13     0     20873.317                         car purchase   
14     1     26420.466          buy residential real estate   

                age_category  
0                      Adult  
1                      Adult  
2                Young adult  
3                Young adult  
4          Middle-aged Adult  
5                Young adult  
6                      Adult  
7          Middle-aged Adult  
8                      Adult  
9                      Adult  
10                     Adult  
11                     Adult  
12  Senior citizen / Elderly  
13         Middle-aged Adult  
14         Middle-aged Adult

In [77]:
df['days_employed'].isna().sum()

2102

`Temos um grande quantidade de valores nulos em ['days_employed']`

In [78]:
df['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

`Valores médios e máximos aqui apresentam problemas, teremos que olhar mais de perto.`

In [79]:
df['dob_years'].max()

75

In [80]:
df['dob_years'].min()

19

In [81]:
df['dob_years'].mean()

43.469025311145295

In [82]:
df['dob_years'].median()

42.0

`Verificando as idades máximas e mínimas dos clientes.`

`Transformando ['days_employed'] em anos para termos uma base.`

In [83]:
df['years_employed'] = df['days_employed'] / 365
df['years_employed']

0         23.116912
1         11.026860
2         15.406637
3         11.300677
4        932.235814
            ...    
21448     12.409087
21449    942.294258
21450      5.789991
21451      8.527347
21452      5.437007
Name: years_employed, Length: 21453, dtype: float64

In [84]:
df['years_employed'].describe()

count    19351.000000
mean       183.328024
std        380.906522
min          0.066141
25%          2.539751
50%          6.011563
75%         15.172281
max       1100.699727
Name: years_employed, dtype: float64

In [85]:
df_impossible_work_time = df[ df['years_employed'] > df['dob_years'] ]
df_impossible_work_time.count()

index               3445
children            3445
days_employed       3445
dob_years           3445
education           3445
education_id        3445
family_status       3445
family_status_id    3445
gender              3445
income_type         3445
debt                3445
total_income        3445
purpose             3445
age_category        3445
years_employed      3445
dtype: int64

`Temos aqui uma enorme quantidade de dados com problemas além dos valores nulos. Onde os anos trabalhados são maiores que a idade do cliente.`

In [86]:
df_impossible_work_time['years_employed'].min()

900.6266317932007

In [87]:
df_impossible_work_time[ ['dob_years','years_employed'] ].head(25)

dob_years  years_employed
4           53      932.235814
18          53     1096.660649
24          57      927.539597
25          67      996.023258
30          62      919.401832
35          68     1079.509787
50          63      969.127212
56          64     1014.096129
71          62      926.338438
78          61      985.542315
86          64     1051.872739
87          53      925.094616
88          63     1022.273396
98          54      999.743029
99          42      949.429093
101         65     1072.764279
104         59     1003.626437
119         60     1047.093504
128         67     1038.454577
129         61      989.069374
133         64      943.855871
136         60      980.493587
142         54      953.281741
150         71     1078.661698
156         65      992.785910

In [88]:
df_years_employed_grouped = df.groupby('age_category')['years_employed'].agg(['mean', 'median'])
df_years_employed_grouped

mean      median
age_category                                    
Adolescent                  1.736104    1.984911
Adult                      29.029981    5.432382
Middle-aged Adult         460.862853   25.072576
Senior citizen / Elderly  860.494599  987.134883
Young adult                 6.986642    3.236526

In [89]:
df_days_employed_grouped = df.groupby('age_category')['days_employed'].agg(['mean', 'median'])
df_days_employed_grouped

mean         median
age_category                                          
Adolescent                   633.678086     724.492610
Adult                      10595.943193    1982.819471
Middle-aged Adult         168214.941448    9151.490125
Senior citizen / Elderly  314080.528722  360304.232308
Young adult                 2550.124368    1181.332061

In [90]:
df_test = df.groupby('dob_years')['years_employed'].agg(['mean', 'median'])
df_test

mean       median
dob_years                         
19           1.736104     1.984911
20           1.876560     1.848874
21           1.943674     1.695161
22           7.652937     1.926877
23           2.266601     1.890970
24           2.812070     2.596523
25           2.981935     2.518354
26           8.807171     2.968926
27          10.207728     3.195102
28           5.972757     3.127960
29           4.257050     3.603982
30           4.646683     3.892019
31           6.386977     3.586031
32          10.622459     3.963349
33           8.880584     3.907986
34          11.345871     4.427151
35           7.548566     4.420534
36          12.319274     4.930193
37          16.107571     4.977297
38          20.935577     4.978614
39          14.458785     5.181886
40          19.387731     4.736813
41          17.536785     5.108651
42          48.607195     6.056372
43          25.803914     5.120602
44          27.265918     5.710493
45          32.186410     6.176524
46          38.688196     5.754721
47          38.969002     6.035832
48          49.725161     6.656643
49          73.095317     7.014567
50         138.337071     7.197223
51         187.467871     7.797481
52         226.858637     9.303122
53         257.734930    10.000021
54         347.281594    11.031620
55         417.164237    19.297181
56         431.280836    18.767615
57         530.654982   914.941073
58         516.634219   907.523224
59         623.279381   943.485856
60         720.343226   957.300102
61         675.975476   942.424471
62         751.729545   972.552653
63         803.791016   982.143277
64         788.138290   974.744007
65         787.530243   978.367034
66         859.215666   996.787504
67         870.218599   978.692754
68         895.607530  1000.762177
69         927.017841   992.639384
70         871.431935   993.939103
71         911.649265   985.482024
72         912.815952  1006.517735
73         859.857992   998.057669
74         660.993586   939.934195
75           4.599917     4.599917

`Trataremos além dos valores nulos, os valores impossíveis.`

In [91]:
# a mediana parece mais segura para organizarmos os dados
df_years_employed_grouped = df.groupby('age_category')['years_employed'].agg(['median'])
df_years_employed_grouped

median
age_category                        
Adolescent                  1.984911
Adult                       5.432382
Middle-aged Adult          25.072576
Senior citizen / Elderly  987.134883
Young adult                 3.236526

`Acima de 55 anos, onde ocorre a maior impossibilidade de anos e dias trabalhados, subtituiremos pelo "Middle-aged Adult"`

In [92]:
median_years = df[ df['age_category'] == 'Middle-aged Adult']['years_employed'].median()
median_years

25.07257568364082

In [93]:
def years_worked(median_years):
    df.loc[(df['years_employed'].isna()) | (df['years_employed'] > median_years), 'years_employed'] = median_years

In [94]:
years_worked(median_years)

In [95]:
df_test = df.groupby('dob_years')['years_employed'].agg(['mean', 'median'])
df_test

mean     median
dob_years                      
19          3.402995   2.030281
20          4.150679   2.190287
21          5.694307   2.050580
22          4.396347   2.170646
23          5.343598   2.253943
24          4.582792   2.734216
25          4.405142   2.618926
26          5.263981   3.257156
27          5.410041   3.567538
28          6.278770   3.763274
29          6.131978   4.164502
30          6.738721   4.492175
31          6.916899   4.413459
32          6.312274   4.170819
33          6.939754   4.449970
34          7.722116   5.029770
35          7.782414   5.352206
36          8.437927   5.980494
37          8.001149   5.528614
38          8.236739   5.795639
39          8.334943   5.892364
40          8.605773   5.851113
41          8.583971   6.084140
42          9.929165   6.878710
43          8.884568   5.899339
44          9.293578   6.453021
45          9.728695   6.925615
46          9.837608   6.641268
47         10.442912   7.860569
48          9.875170   7.272132
49         10.911026   7.916975
50         11.823269   8.423671
51         12.614571   8.934461
52         13.599401  11.154370
53         13.870272  11.804374
54         14.988081  15.716635
55         16.802929  25.072576
56         16.543482  25.072576
57         18.148804  25.072576
58         17.837384  25.072576
59         19.622088  25.072576
60         21.117359  25.072576
61         20.700558  25.072576
62         21.178045  25.072576
63         22.541420  25.072576
64         22.118703  25.072576
65         22.108690  25.072576
66         23.107352  25.072576
67         23.203041  25.072576
68         23.644683  25.072576
69         24.032979  25.072576
70         23.359009  25.072576
71         23.512735  25.072576
72         24.359825  25.072576
73         23.112889  25.072576
74         20.556375  25.072576
75          4.599917   4.599917

In [96]:
median_days = df[ df['age_category'] == 'Middle-aged Adult']['days_employed'].median()
median_days

9151.4901245289

In [97]:
def days_worked(median_days):
    df.loc[(df['days_employed'].isna()) | (df['days_employed'] > median_days), 'days_employed'] = median_days

In [98]:
days_worked(median_days)

In [99]:
df_test = df.groupby('dob_years')['days_employed'].agg(['mean', 'median'])
df_test

mean       median
dob_years                          
19         1242.093232   741.052617
20         1514.997820   799.454693
21         2078.421880   748.461630
22         1604.666613   792.285881
23         1950.413180   822.689270
24         1672.719036   997.988977
25         1607.876829   955.907868
26         1921.352946  1188.862091
27         1974.664897  1302.151470
28         2291.751186  1373.594888
29         2238.171875  1520.043109
30         2459.633009  1639.643796
31         2524.668190  1610.912684
32         2303.979837  1522.348812
33         2533.010178  1624.239059
34         2818.572307  1835.865868
35         2840.581198  1953.555101
36         3079.843506  2182.880449
37         2920.419235  2017.943999
38         3006.409794  2115.408106
39         3042.254292  2150.712995
40         3141.106991  2135.656248
41         3133.149311  2220.710993
42         3624.145217  2510.729122
43         3242.867339  2153.258732
44         3392.155810  2355.352766
45         3550.973570  2527.849297
46         3590.726813  2424.062686
47         3811.662804  2869.107543
48         3604.437168  2654.328274
49         3982.524449  2889.695878
50         4315.493037  3074.639866
51         4604.318350  3261.078178
52         4963.781344  4071.345128
53         5062.649142  4308.596384
54         5470.649535  5736.571825
55         6133.069043  9151.490125
56         6038.371054  9151.490125
57         6624.313617  9151.490125
58         6510.645331  9151.490125
59         7162.062226  9151.490125
60         7707.836143  9151.490125
61         7555.703728  9151.490125
62         7729.986564  9151.490125
63         8227.618438  9151.490125
64         8073.326564  9151.490125
65         8069.672004  9151.490125
66         8434.183401  9151.490125
67         8469.109904  9151.490125
68         8630.309356  9151.490125
69         8772.037394  9151.490125
70         8526.038165  9151.490125
71         8582.148401  9151.490125
72         8891.336229  9151.490125
73         8436.204545  9151.490125
74         7503.076807  9151.490125
75         1678.969771  1678.969771

In [100]:
# Verifique as entradas em todas as colunas - certifique-se de corrigir todos os valores ausentes

In [101]:
df.isna().sum()

index               0
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_category        0
years_employed      0
dtype: int64

`Problemas corrigidos, seguimos em frente para a categorização dos dados.`

In [102]:
# analisando children
df['children'].unique()

array([1, 0, 3, 2, 4, 5])

In [103]:
# analisando debt
df['debt'].unique()

array([0, 1])

In [104]:
# criando coluna ['financial_status'] 
def financial_status(df):
    df.loc[df['debt'] == 0, 'financial_status'] = 'creditor'
    df.loc[df['debt'] == 1, 'financial_status'] = 'debtor'
    return df

In [105]:
df = financial_status(df)

In [106]:
df['financial_status'].describe()

count        21453
unique           2
top       creditor
freq         19712
Name: financial_status, dtype: object

`Criamos uma nova categoria para as próximas análises.`

In [107]:
# analisando o número de filhos
df_children_grouped = df.groupby(['children', 'financial_status']).count()
df_children_grouped

index  days_employed  dob_years  education  \
children financial_status                                               
0        creditor          13027          13027      13027      13027   
         debtor             1063           1063       1063       1063   
1        creditor           4410           4410       4410       4410   
         debtor              445            445        445        445   
2        creditor           1926           1926       1926       1926   
         debtor              202            202        202        202   
3        creditor            303            303        303        303   
         debtor               27             27         27         27   
4        creditor             37             37         37         37   
         debtor                4              4          4          4   
5        creditor              9              9          9          9   

                           education_id  family_status  family_status_id  \
children financial_status                                                  
0        creditor                 13027          13027             13027   
         debtor                    1063           1063              1063   
1        creditor                  4410           4410              4410   
         debtor                     445            445               445   
2        creditor                  1926           1926              1926   
         debtor                     202            202               202   
3        creditor                   303            303               303   
         debtor                      27             27                27   
4        creditor                    37             37                37   
         debtor                       4              4                 4   
5        creditor                     9              9                 9   

                           gender  income_type   debt  total_income  purpose  \
children financial_status                                                      
0        creditor           13027        13027  13027         13027    13027   
         debtor              1063         1063   1063          1063     1063   
1        creditor            4410         4410   4410          4410     4410   
         debtor               445          445    445           445      445   
2        creditor            1926         1926   1926          1926     1926   
         debtor               202          202    202           202      202   
3        creditor             303          303    303           303      303   
         debtor                27           27     27            27       27   
4        creditor              37           37     37            37       37   
         debtor                 4            4      4             4        4   
5        creditor               9            9      9             9        9   

                           age_category  years_employed  
children financial_status                                
0        creditor                 13027           13027  
         debtor                    1063            1063  
1        creditor                  4410            4410  
         debtor                     445             445  
2        creditor                  1926            1926  
         debtor                     202             202  
3        creditor                   303             303  
         debtor                      27              27  
4        creditor                    37              37  
         debtor                       4               4  
5        creditor                     9               9

In [108]:
# categorizando por número de filhos
def children_status(df):
    df.loc[df['children'] == 0, 'children_status'] = 'no child'
    df.loc[df['children'] >= 3, 'children_status'] = '3 or more children'
    df.loc[(df['children'] > 0) & (df['children'] < 3), 'children_status'] = '1-2 children'
    
    return df

In [109]:
df = children_status(df)
df

index  children  days_employed  dob_years            education  \
0          0         1    8437.673028         42    bachelor's degree   
1          1         1    4024.803754         36  secondary education   
2          2         0    5623.422610         33  secondary education   
3          3         3    4124.747207         32  secondary education   
4          4         0    9151.490125         53  secondary education   
...      ...       ...            ...        ...                  ...   
21448  21520         1    4529.316663         43  secondary education   
21449  21521         0    9151.490125         67  secondary education   
21450  21522         1    2113.346888         38  secondary education   
21451  21523         3    3112.481705         38  secondary education   
21452  21524         2    1984.507589         40  secondary education   

       education_id      family_status  family_status_id gender income_type  \
0                 0            married                 0      F    employee   
1                 1            married                 0      F    employee   
2                 1            married                 0      M    employee   
3                 1            married                 0      M    employee   
4                 1  civil partnership                 1      F     retiree   
...             ...                ...               ...    ...         ...   
21448             1  civil partnership                 1      F    business   
21449             1            married                 0      F     retiree   
21450             1  civil partnership                 1      M    employee   
21451             1            married                 0      M    employee   
21452             1            married                 0      F    employee   

       debt  total_income                  purpose              age_category  \
0         0     40620.102    purchase of the house                     Adult   
1         0     17932.802             car purchase                     Adult   
2         0     23341.752    purchase of the house               Young adult   
3         0     42820.568  supplementary education               Young adult   
4         0     25378.572        to have a wedding         Middle-aged Adult   
...     ...           ...                      ...                       ...   
21448     0     35966.698     housing transactions                     Adult   
21449     0     24959.969        purchase of a car  Senior citizen / Elderly   
21450     1     14347.610                 property                     Adult   
21451     1     39054.888        buying my own car                     Adult   
21452     0     13127.587             to buy a car                     Adult   

       years_employed financial_status     children_status  
0           23.116912         creditor        1-2 children  
1           11.026860         creditor        1-2 children  
2           15.406637         creditor            no child  
3           11.300677         creditor  3 or more children  
4           25.072576         creditor            no child  
...               ...              ...                 ...  
21448       12.409087         creditor        1-2 children  
21449       25.072576         creditor            no child  
21450        5.789991           debtor        1-2 children  
21451        8.527347           debtor  3 or more children  
21452        5.437007         creditor        1-2 children  

[21453 rows x 17 columns]

In [110]:
# avaliando a distribuição da renda
df['total_income'].describe()

count     21453.000000
mean      26436.621167
std       15703.396512
min        3306.762000
25%       17219.352000
50%       22711.042000
75%       31331.009000
max      362496.645000
Name: total_income, dtype: float64

In [111]:
# criando categorias de renda
def income_class(df):
    df.loc[df['total_income'] >= 40000, 'income_class'] = 'high class'
    df.loc[df['total_income'] <= 20000, 'income_class'] = 'low class'
    df.loc[(df['total_income'] > 20000) & (df['total_income'] < 40000), 'income_class'] = 'middle class'
    
    return df

In [112]:
df = income_class(df)
df_test = df.groupby(['income_class', 'financial_status']).count()
df_test

index  children  days_employed  dob_years  \
income_class financial_status                                              
high class   creditor           2618      2618           2618       2618   
             debtor              194       194            194        194   
low class    creditor           6761      6761           6761       6761   
             debtor              608       608            608        608   
middle class creditor          10333     10333          10333      10333   
             debtor              939       939            939        939   

                               education  education_id  family_status  \
income_class financial_status                                           
high class   creditor               2618          2618           2618   
             debtor                  194           194            194   
low class    creditor               6761          6761           6761   
             debtor                  608           608            608   
middle class creditor              10333         10333          10333   
             debtor                  939           939            939   

                               family_status_id  gender  income_type   debt  \
income_class financial_status                                                 
high class   creditor                      2618    2618         2618   2618   
             debtor                         194     194          194    194   
low class    creditor                      6761    6761         6761   6761   
             debtor                         608     608          608    608   
middle class creditor                     10333   10333        10333  10333   
             debtor                         939     939          939    939   

                               total_income  purpose  age_category  \
income_class financial_status                                        
high class   creditor                  2618     2618          2618   
             debtor                     194      194           194   
low class    creditor                  6761     6761          6761   
             debtor                     608      608           608   
middle class creditor                 10333    10333         10333   
             debtor                     939      939           939   

                               years_employed  children_status  
income_class financial_status                                   
high class   creditor                    2618             2618  
             debtor                       194              194  
low class    creditor                    6761             6761  
             debtor                       608              608  
middle class creditor                   10333            10333  
             debtor                       939              939

In [113]:
# examinando a coluna purpose
df['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [114]:
# contando os propósitos
len(df['purpose'].unique())

38

In [115]:
# assimilando propósitos house
df['purpose'] = df['purpose'].replace(['purchase of the house', 'housing transactions', 'buy real estate',
                'buy commercial real estate', 'buy residential real estate',
                'construction of own property', 'property', 'building a property',
                'housing', 'transactions with my real estate', 'purchase of my own house',
                'real estate transactions', 'buying property for renting out',
                'housing renovation','transactions with commercial real estate', 'building a real estate', 'purchase of the house for my family'], 'buy, build or renovate a house')

In [116]:
# assimilando propósitos education
df['purpose'] = df['purpose'].replace(['supplementary education', 'education', 'to become educated',
                 'getting an education', 'profile education', 'university education',
                 'to get a supplementary education', 'getting higher education', 'going to university'], 'investment in education')

In [117]:
# assimilando propósitos wedding
df['purpose'] = df['purpose'].replace(['to have a wedding', 'having a wedding', 'wedding ceremony'], 'wedding')

In [118]:
# assimilando propósitos car
df['purpose'] = df['purpose'].replace(['car purchase', 'buying a second-hand car', 'buying my own car',
                   'cars', 'second-hand car purchase', 'car', 'to own a car', 'purchase of a car',
                   'to buy a car'], 'buy a car')

In [119]:
df['purpose'].unique()

array(['buy, build or renovate a house', 'buy a car',
       'investment in education', 'wedding'], dtype=object)

In [120]:
len(df['purpose'].unique())

4

`Tudo pronto, vamos testar as hipóteses.`

## Verificar as Hipóteses


**Existe uma correlação entre o nível de renda e do pagamento em dia?**

In [121]:
df.columns

Index(['index', 'children', 'days_employed', 'dob_years', 'education',
       'education_id', 'family_status', 'family_status_id', 'gender',
       'income_type', 'debt', 'total_income', 'purpose', 'age_category',
       'years_employed', 'financial_status', 'children_status',
       'income_class'],
      dtype='object')

In [122]:
# Verifique os dados de renda e do pagamento em dia
income_x_payment = df.pivot_table(index='income_class', values='debt', aggfunc='sum')
income_x_payment

debt
income_class      
high class     194
low class      608
middle class   939

In [123]:
# taxa de inadimplência com base na renda
default_rate_income =  (income_x_payment / len(df)) * 100
default_rate_income

debt
income_class          
high class    0.904302
low class     2.834102
middle class  4.377010

`Analisando as classes de renda, podemos considerar que a 'high class' dificilmente atrasa os pagamentos, o percentual de debtor é abaixo de 1%. Já para as classes 'low' e 'middle' são entre 2% e chegando a quase 5%.`

`Podemos então concluir que: baseando na classe do cliente, podemos estimar a tendência de atraso nos pagamentos.`

In [124]:
# Verifique os dados de criança e do pagamento em dia
children_x_payment = df.pivot_table(index='children_status', values='debt', aggfunc='sum')
children_x_payment

debt
children_status         
1-2 children         647
3 or more children    31
no child            1063

In [125]:
# Calcular a taxa de inadimplência com base no número de filhos
default_rate_children =  (children_x_payment / len(df)) * 100
default_rate_children

debt
children_status             
1-2 children        3.015895
3 or more children  0.144502
no child            4.955018

`Temos aqui dados interessantes!`

`A taxa de inadimplência vai decrescendo com a maior quantidade de filhos. Quanto mais filhos, mais pagam em dia.`

`Pessoas sem filhos são as menos confiáveis.`

**Existe uma correlação entre o status familiar e o pagamento em dia?**

In [126]:
# Verifique os dados de status da família e do pagamento em dia
family_x_payment = df.pivot_table(index='family_status', values='debt', aggfunc='sum')
family_x_payment

debt
family_status          
civil partnership   388
divorced             85
married             931
unmarried           274
widow / widower      63

In [127]:
# Calcular a taxa padrão com base no status da família
default_rate_family =  (family_x_payment / len(df)) * 100
default_rate_family

debt
family_status              
civil partnership  1.808605
divorced           0.396215
married            4.339719
unmarried          1.277211
widow / widower    0.293665

`Com base no status familiar, os percentuais variam bastante. Mas sempre é uma boa ideia emprestar para os divorciados e viúvos.`

**Como a finalidade do crédito afeta a taxa de inadimplência?**

In [128]:
# Confira os percentuais de inadimplência para cada finalidade de crédito e analise-os
purpose_x_payment = df.pivot_table(index='purpose', values='debt', aggfunc='sum')
purpose_x_payment

debt
purpose                             
buy a car                        403
buy, build or renovate a house   782
investment in education          370
wedding                          186

In [129]:
default_rate_purpose =  (purpose_x_payment / len(df)) * 100
default_rate_purpose

debt
purpose                                 
buy a car                       1.878525
buy, build or renovate a house  3.645178
investment in education         1.724701
wedding                         0.867012

`Investir em uma casa sempre é mais caro e o risco de inadimplência é maior. Vamos tomar cuidado!`

`Aqueles clientes que querem fazer uma festa de casamento ou investir em educação tem mais chances de pagar os saldos.`

`Este projeto 2 foi de enorme aprendizado e um desafio constante para fazer essa análise.`

`Devido ao dataset estar com valores nulos e confusos, foi preciso limpar e organizar bem para ter uma visão geral.`

`Dados de dias trabalhados estavam por algum motivo no qual não sabemos, negativos. Fizemos os ajustes necessários.`

`Faixa de renda tb oscilava muito, então agrupamos em high, low and middle.`

`Tivemos que categorizar e colocar alguns rótulos nos clientes, além de agrupá-los em alguns momentos. Principalmente em relação aos propósitos, onde eram similares.`

`Percebemos que os clientes sem filhos são os mais relapsos quanto ao pagamento em dia. Quanto mais filhos, melhores pagadores.`

`Faixa de renda também altera e influencia no pagamento, quanto maior a renda do cliente, maiores as chances de pagamento. Isso segue uma normalidade prática.`

`Já no status familiar, mostrou-se bem interessante, onde os divorciados e viúvos, são os melhores clientes.`

`Com base no propósito para analisar o crédito, tivemos um resultado bem interessante. Se tivéssemos os valores dos empréstimos adquiridos ou requeridos pelos clientes, poderíamos estimar melhor. Mostrou-se que aqueles que pediram valores para propósitos de 'home' são mais inadimplentes dos que pediram para estudos ou fazer a festa de casamento.`